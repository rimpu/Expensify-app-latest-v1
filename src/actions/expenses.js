import uuid from 'uuid';
import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = (expense) => {

  return {
  type: 'ADD_EXPENSE',
  expense
}};

export const startAddExpense = (expenseData ={})=>{
  return (dispatch,getState)=>{
    const { description = '', note = '', amount = 0, createdAt = 0} = {...expenseData}
    const expense = expenseData
    console.log(getState())
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot)=>{
      const expense1 = {id: snapshot.key,...expense}
      dispatch(addExpense(expense1))
    })

  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense =({id}={})=>{
  return(dispatch,getState)=>{
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
      dispatch(removeExpense({id}))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpenses = (id,updates)=>{
  return (dispatch,getState)=>{
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(()=>{
      dispatch(editExpense(id,updates))
    })
  }
}

//SET_EXPENSES

export const setExpenses =(expenses)=>{
  return {
  type:'SET_Expenses',
  expenses
}}

export const startSetExpenses =() =>{
  let expenses=[]
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value',(snapshot)=>{
    snapshot.forEach(element => {
      expenses.push({id: element.key,...element.val()})
    });
  }).then(()=>{
    dispatch(setExpenses(expenses))
  })
}
}


