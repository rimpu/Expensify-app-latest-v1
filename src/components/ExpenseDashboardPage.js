import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummaryConnected from '../components/ExpensesSummary'


const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummaryConnected />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
