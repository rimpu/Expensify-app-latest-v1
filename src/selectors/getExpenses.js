export const getExpenses =(expenses)=>{
    let expensesAmount = expenses.map((expense)=> { return expense.amount})
    let total = expensesAmount.reduce((accumulator, currentValue) => accumulator + currentValue,0 )
    return total
}