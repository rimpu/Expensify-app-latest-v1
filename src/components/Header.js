import React from 'react';
import { NavLink } from 'react-router-dom';
import {connect} from 'react-redux'
import {LogOutUser} from '../actions/auth';

export const Header = (props) => {
  console.log(props)
  return (
    <header>
    <h1>Expensify</h1>
    <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    <button onClick={props.LogOutUser} >Log out</button>
    </header>
  )
 
  };

const mapDispatchToProps =(dispatch)=>{
  return {
    LogOutUser : ()=> dispatch(LogOutUser())
  }
}

export default connect(undefined,mapDispatchToProps)(Header);
