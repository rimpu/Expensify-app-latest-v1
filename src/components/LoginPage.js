import React from 'react'
import {connect} from 'react-redux'
import {AuthoriseUser} from '../actions/auth'

export const LoginButton = ({ AuthoriseUser }) => (
    <div>
      <button onClick={AuthoriseUser}>Login</button>
    </div>
  );
  
  const mapDispatchToProps = (dispatch) => ({
    AuthoriseUser: () => dispatch(AuthoriseUser())
  });

export default connect(undefined,mapDispatchToProps)(LoginButton)