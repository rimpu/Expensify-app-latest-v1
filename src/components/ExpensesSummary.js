import React from 'react'
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses'
import {getExpenses} from '../selectors/getExpenses'
import numeral from 'numeral'

export class ExpensesSummary extends React.Component{
    render(){
        return(
            <p>Viewing {this.props.expenses.length} expense{this.props.expenses.length===1 ? '' :'s'} totalling {numeral(this.props.total / 100).format('$0,0.00')}</p>
        )
    }
} 

const mapStateToProps =(state)=>{
   return {
      expenses: selectExpenses(state.expenses, {...state.filters}),
      total : getExpenses(selectExpenses(state.expenses, {...state.filters}))
    }
}

export default connect(mapStateToProps)(ExpensesSummary)