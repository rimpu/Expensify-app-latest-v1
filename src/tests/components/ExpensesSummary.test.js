import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'
import {getExpenses} from '../../selectors/getExpenses'

test('Should render 0 expenses summary',()=>{
    const wrapper = shallow(<ExpensesSummary expenses={[]} total={getExpenses([])}/>)
    expect(wrapper).toMatchSnapshot()
})

test('Should render multiple expenses summary',()=>{
    const wrapper = shallow(<ExpensesSummary expenses={expenses} total={getExpenses(expenses)}/>)
    expect(wrapper).toMatchSnapshot()
})