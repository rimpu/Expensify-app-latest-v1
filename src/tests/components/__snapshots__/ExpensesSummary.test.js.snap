// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render 0 expenses summary 1`] = `
<p>
  Viewing 
  0
   expense
  s
   totalling 
  $0.00
</p>
`;

exports[`Should render multiple expenses summary 1`] = `
<p>
  Viewing 
  3
   expense
  s
   totalling 
  $1,141.95
</p>
`;
